lib/streamlit/report_session.py

```diff
@@ -86,6 +87,7 @@ def __init__(self, ioloop, script_path, command_line, uploaded_file_manager):
         self._ioloop = ioloop
         self._report = Report(script_path, command_line)
         self._uploaded_file_mgr = uploaded_file_manager
+        self._repo = GitRepo(self._report.script_path)
```

Comment:
Is ReportSession.py the right place to initialize this, instead of inside Server.py? According to our server diagram: https://www.notion.so/streamlit/Eng-docs-3d5077d3b0524a62bb375b76f0fe4552

It looks like ReportSession is instantiated once per websocket connection (each new browser tab); but the GitRepo parsing should only need to happen once on `streamlit run`. After all, the location of script_path isn't changing once per browser tab.

I expect the GitRepo initialization to be pretty fast -- it should just be reading files on local disk -- but in case it's not (disk is getting clobbered, user is on an old spinning HDD instead of an SSD), we'd rather do this exactly once.

(Actually, if we're putting GitRepo into the critical path, it'd be worth verifying how fast this thing runs with a quick timing session. If it adds significant (>0.1s?) startup latency, we might want to background this and make it run asynchronously somehow. I *don't* think this is the case fwiw, but it should be quick to check with timeit or literally just logging and looking at the timestamp diffs)

Reply:
Hmmm...depends on a couple things, assuming the tab is open, the server can change under the hood. Two examples now that I am thinking about it.

The user closes and opens a new file via `streamlit run` this will issue a reconnection. My other idea was the `initialize` message. I thought creating it on the report session was more meaningful because the script's path is there to be sent down thinking it's probably more relevant especially if the file changes under the hood.

The other option I should test is if the user switches branches. The file may change and issue a rerun. That rerun would then update the branch. I'm actually thinking this use case might not work, but I should bring it up as a discussion point.

Reply:
For the first case, if a user opens a new file via `streamlit run` doesn't that mean a new server is being spun up? And it's okay if the file contents change under the hood, as long as the repo/branch/filename don't change, right?

Hadn't thought about the second case though -- good point, switching branches would mean we're filling in the wrong info if they reconnect. I suspect the "switching branches" use case would still give the old branch if they rerun (which is in ScriptRunner.py by the diagram), or if they just switch branches and then go back and deploy; so even putting it in ReportSession is not foolproof.

With my product hat on, I'd say switching branches while `streamlit run` is running is a minor use case anyways. The important thing is probably, "when the user clicks the deploy button, the version of the app we deploy is the one they see in front of them". That might be an argument in favor of inside ScriptRunner (???)

Ultimately, the call is up to you, just wanted to check that we deliberately thought about GitRepo's placement and performance in the context of the Streamlit lifecycle.

Reply:
I'll bring it up with Product at Standup

Reply:
According to @tvst , the New Report is actually run every rerun, so it's the preferred solution. Going to test the branch switching to make sure that works.

Reply:
Tested it out, and changing branches _does_ work!